draft-connolly-cfrg-xwing-kem.md

```diff
@@ -219,220 +239,233 @@ operations, roles, and behaviors of HPKE:
-    its encapsulated representation `enc`. This function can raise a
-    `DecapError` on decapsulation failure.
-  - `ML-KEM-768.AuthEncap(pkR, skS)` (optional): Same as `Encap()`, and the outputs
-    encode an assurance that the KEM shared secret was generated by the
-    holder of the private key `skS`.
-  - `ML-KEM-768.AuthDecap(enc, skR, pkS)` (optional): Same as `Decap()`, and the recipient
-    is assured that the KEM shared secret was generated by the holder of
-    the private key `skS`.
-  - `ML-KEM-768.Nsecret`: The length in bytes of a KEM shared secret produced by this KEM.
-  - `ML-KEM-768.Nenc`: The length in bytes of an encapsulated key produced by this KEM.
-  - `ML-KEM-768.Npk`: The length in bytes of an encoded public key for this KEM.
-  - `ML-KEM-768.Nsk`: The length in bytes of an encoded private key for this KEM.
+
+* ML-KEM-768 post-quantum key-encapsulation mechanism (KEM), TODO ref:
+
+  - `ML-KEM-768.DeriveKeyPair(ikm)`: Deterministic algorithm to
+    derive an ML-KEM-768 key pair `(sk_M, pk_M)` from entropy ikm.
+  - `ML-KEM-768.Encapsulate(pk_M)`: Randomized algorithm to generate (ss_M, ct_M),
+    an ephemeral 32 byte shared key ss_M,
+    and a fixed-length encapsulation (cipher text) of that key ct_M for pk_M.
+  - `ML-KEM-768.Decap(ct_M, sk_M)`: Deterministic algorithm using the
+    private key `sk_M` to recover the shared key
+    from ct_M.
 
 * X25519 elliptic curve Diffie-Hellman key-exchange defined in {{Section 5 of RFC7748}}:
 
-  - `X25519.GenerateKeyPair()`: Randomized algorithm to generate an X25519 key pair `(skX, pkX)`
-  - `X25519.DeriveKeyPair(ikm)`: Deterministic algorithm to derive a key pair `(skX,
-    pkX)` from the byte string `ikm`, where `ikm` SHOULD have at least `Nsk`
-    bytes of entropy (see {{derive-key-pair}} for discussion).
-  - `X25519.SerializePublicKey(pkX)`: Produce a byte string of length `Npk` encoding
-    the public key `pkX`.
-  - `X25519.DeserializePublicKey(pkXm)`: Parse a byte string of length `Npk` to recover a
-    public key. This function can raise a `DeserializeError` error upon `pkXm`
-    deserialization failure.
-    TODO: lock this down to be DH based on the base point, not just generic X25519 function
-  - `X25519.DH(skX, pkY)`: Perform a non-interactive Diffie-Hellman exchange over
-    the Montgomery form of the elliptic curve curve25519 using the private key
-    `skX` and public key `pkY` to produce a Diffie-Hellman shared secret of
-    length `Ndh` as defined in {{Section 5 of RFC7748}}. This function can raise
-    a `ValidationError` as described in {{validation}}.
-  - `X25519.Ndh`: The length in bytes of a Diffie-Hellman shared secret produced by
-    `X25519()`, which is 32.
-  - `X25519.Nsk`: The length in bytes of a Diffie-Hellman private key, which is 32.
-
-* Hash functions:
-
-  - `SHAKE128(bytes)`:
-  - `SHA3-256(bytes)`:
+  - `X25519(k,u)`: takes 32 byte strings k and u representing a
```

Comment:
We need to define the keygen algorithm and point at RFC 7748

Reply:
And we need the component scheme serialization routines

Reply:
As in the paper since we need keygen() and serialization stuff having our definition of `X25519 { keygen(), DH(), ... }` etc makes this nicer

Reply:
I don't think we need to: I inline the keygen into the definition of X-Kem. That also fits with the proof.

Reply:
Also this makes it simpler to implement as most libraries only expose the DH function of X25519.

Reply:
(eg. https://pkg.go.dev/golang.org/x/crypto/curve25519 )

Reply:
> Also this makes it simpler to implement as most libraries only expose the DH function of X25519.

hm, fair

Reply:
I removed the ugly hardcoded 9.